### PR TITLE
feat: Third Person Character Controller integration with custom IP UI

### PR DESCRIPTION
**Pull request by fernando-cortez**
This PR adds a new Player Prefab for NetworkManager, based off of the Third Person Controller Starter Assets. Previous functionality of movement, collisions, and picking up is intact. The new Input System has been integration for input polling. ClientNetworkAnimator enables the animations of the player to be client-driven.

Included as well is an overhaul on the UI to now include custom IPs, using UI Toolkit.

[MTT-5013](https://jira.unity3d.com/browse/MTT-5013)

Package Manager manifest will target the new Boss Room release containing the new Utilities release, once that is out!

Testing instructions:
* Create a project clone via ParrelSync, or create a build of the game. To test the game via the editor, open up the Bootstrap scene and hit Play.
* Have one instance of the game host, and have other instances join the host.
* Here, validate that all clients see animations being played out from every other client. Test picking up ingredients with E.

### CODE REVIEW

File: Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs
Comment (LPLafontaineB) [inline Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:44]: Do we need to set it false here too since it is already set to false on awake?

File: Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs
Comment (LPLafontaineB) [inline Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:64]: should we do this todo now while we are here?

File: Basic/ClientDriven/Assets/Scripts/HostJoinUI.cs
Comment (LPLafontaineB) [inline Basic/ClientDriven/Assets/Scripts/HostJoinUI.cs:55]:
```suggestion
        var result = NetworkManager.Singleton.StartHost();
```
Same thing in the other methods below

**Review by LPLafontaineB — COMMENTED** [review]
Mostly minor nitpicks in code. Otherwise I see an issue where when we drop an object it disappears into the ground. Also another really nitpicky comment, but since we can now jump, maybe the "counter" should be a bit higher so that we don't notice the invisible wall too much? It might be something for a later PR though

**Comment by RikuTheFuffs** [comment]
@fernando-cortez would it be possible to have some testing instructions listed in the PR? I'd like to test it locally, but dont't really know if I should do something in particular to set up the environment or if there's a specific key combiantion to test

**Comment by fernando-cortez** [comment]
> @fernando-cortez would it be possible to have some testing instructions listed in the PR? I'd like to test it locally, but dont't really know if I should do something in particular to set up the environment or if there's a specific key combination to test

Absolutely, I'll include some test instructions in the summary!

File: Basic/ClientDriven/Assets/Scripts/ServerPlayerMove.cs
Comment (RikuTheFuffs) [inline Basic/ClientDriven/Assets/Scripts/ServerPlayerMove.cs:17]: as per our [coding conventions](https://github.cds.internal.unity3d.com/unity/unity-meta/blob/stable/ReferenceSource/CSharp/Assets/CSharpReference.cs), we should not use abbreviations in code, unless we're dealing with acronyms

File: Basic/ClientDriven/Assets/Textures/backgroundGlow.png.meta
Comment (RikuTheFuffs) [inline Basic/ClientDriven/Assets/Textures/backgroundGlow.png.meta:9]: As this is used in a screen-overlay UI mipmaps should be disabled (saves an amount of memory equal to ~33% of the texture's weight)

**Review by RikuTheFuffs — CHANGES_REQUESTED** [review]
1. There seem to be some warnings related to the Chintzy font (no stack trace). Do you know what's causing them? Did you try addressing them already? You can let them pop up by dropping a colored object in one of the circles of the same color as the object

image

2. There's something wrong with one of the network variables:

image

3. Going out of play mode after playing for a while thows:


```
Type of instance in array does not match expected type
UnityEngine.Object:Destroy (UnityEngine.Object)
Unity.Netcode.NetworkSpawnManager:OnDespawnObject (Unity.Netcode.NetworkObject,bool) (at Library/PackageCache/com.unity.netcode.gameobjects@1.1.0/Runtime/Spawning/NetworkSpawnManager.cs:873)
Unity.Netcode.NetworkSpawnManager:DespawnAndDestroyNetworkObjects () (at Library/PackageCache/com.unity.netcode.gameobjects@1.1.0/Runtime/Spawning/NetworkSpawnManager.cs:699)
Unity.Netcode.NetworkManager:ShutdownInternal () (at Library/PackageCache/com.unity.netcode.gameobjects@1.1.0/Runtime/Core/NetworkManager.cs:1439)
Unity.Netcode.NetworkManager:OnDestroy () (at Library/PackageCache/com.unity.netcode.gameobjects@1.1.0/Runtime/Core/NetworkManager.cs:1331)
Unity.Netcode.NetworkManager:OnApplicationQuit () (at Library/PackageCache/com.unity.netcode.gameobjects@1.1.0/Runtime/Core/NetworkManager.cs:1325)

```

**Comment by RikuTheFuffs** [comment]
Also, I'm not sure if this is on purpose, but there seem to be some interactivity problems:

1. Dropping an object outside of a circle by pressingthe E button again makes the object fall under the ground
2. Under certain circumstances, I was able to grab objects very far away and to make them follow my character's rotation. I didn't manage to repro this consistently, but spamming the E button probably helps
3. Sometimes, I was able to pick up objects even when I had nothing around me. Pressing E would just make something pop up on me, as if I summoned it out of thin air

File: Basic/ClientDriven/Assets/Scripts/HostJoinUI.cs
Comment (fernando-cortez) [reply Basic/ClientDriven/Assets/Scripts/HostJoinUI.cs:107]: Would you be able to direct me to the specific bit where it specifies this? We've used vars even without context inside of Boss Room so I'm curious about this.

**Comment by fernando-cortez** [comment]
@RikuTheFuffs addressed the comments you've raised and left feedback on others as well. The interactivity of the sample will feel a lot better, with non-allocating boxcasts from the players direction now determining whether an object should be picked up. Picked up objects also now don't fall through the floor 😅

File: Basic/ClientDriven/Assets/Scripts/HostJoinUI.cs
Comment (RikuTheFuffs) [reply Basic/ClientDriven/Assets/Scripts/HostJoinUI.cs:107]: Oh, you actually made me realize that the documentation says (at [this line](https://github.cds.internal.unity3d.com/unity/unity-meta/blob/c0eeb02f04bdd06f0aedc0eee78e67ce20786c5d/ReferenceSource/CSharp/Assets/CSharpReference.cs#L223)) that you should use var when the type is explicit, but says nothing about what to do when it is not. My bad! I guess what's written here is fine, then.

File: Basic/ClientDriven/Assets/Scripts/ServerIngredient.cs
Comment (RikuTheFuffs) [inline Basic/ClientDriven/Assets/Scripts/ServerIngredient.cs:20]: public fields are automatically serialized, so the SerializeField here is redundant (or even dangerous if it causes the serialization o happen twice). I'd suggest to either use `internal` + `SerializeField`, or to drop the `SerializeField`

File: Basic/ClientDriven/Assets/Scripts/ServerScoreReplicator.cs
Comment (RikuTheFuffs) [inline Basic/ClientDriven/Assets/Scripts/ServerScoreReplicator.cs:7]: Nitpick: being a property, it should be `ReplicatedScore` (link [here](https://github.cds.internal.unity3d.com/unity/unity-meta/blob/c0eeb02f04bdd06f0aedc0eee78e67ce20786c5d/ReferenceSource/CSharp/Assets/CSharpReference.cs#L194))

**Review by RikuTheFuffs — APPROVED** [review]
I just retested it and it looks great! Apart from the `[SerializeField]` -related note i left, everything looks good from my side

File: Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs
Comment (SamuelBellomo) [inline Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:1]: Adding a note here, not linked to this file.
wanna add an entry to the changelog please?

File: Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs
Comment (SamuelBellomo) [inline Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:56]: Was the doc updated?

File: Basic/ClientDriven/Assets/Scripts/ServerIngredient.cs
Comment (SamuelBellomo) [reply Basic/ClientDriven/Assets/Scripts/ServerIngredient.cs:20]: I'd be surprised if public + SerializeField cause the serialization to happen twice, any link to that?

File: Basic/ClientDriven/Assets/Scripts/ServerIngredientSpawner.cs
Comment (SamuelBellomo) [inline Basic/ClientDriven/Assets/Scripts/ServerIngredientSpawner.cs:53]: I don't see spawnWaves being reset anywhere or decremented. Does it mean if players place all ingredients in the circles, there won't be new ingredients?
(That can work, I don't mind, just checking if this is the intended behaviour)

File: Basic/ClientDriven/Assets/Scripts/ServerPlayerMove.cs
Comment (SamuelBellomo) [reply Basic/ClientDriven/Assets/Scripts/ServerPlayerMove.cs:28]: inherit doc is for xml doc generation. The above is a note for developers, telling us to go to our docs website and update the line numbers that reference the samples. There's (unfortunately) no automation with the above "DOC START HERE" :(

File: Basic/ClientDriven/Assets/Scripts/ServerPlayerMove.cs
Comment (SamuelBellomo) [inline Basic/ClientDriven/Assets/Scripts/ServerPlayerMove.cs:97]: When we were looking at this, I thought this didn't fix the issue, was that it in the end? Or did you forget to reverse that change?

Comment (SamuelBellomo) [reply Basic/ClientDriven/Assets/Scripts/ServerPlayerMove.cs:97]: I guess because RPCs can happen in update, deferring physics changes to FixedUpdate makes sense still... nvm the above comment.
I'm still curious though whether that fixed the issue or not

Comment (SamuelBellomo) [reply Basic/ClientDriven/Assets/Scripts/ServerPlayerMove.cs:97]: If it did, it'd be great to comment this and produce some edu comments.

File: Basic/ClientDriven/Assets/StarterAssets/ThirdPersonController/Scripts/ThirdPersonController.cs
Comment (SamuelBellomo) [reply Basic/ClientDriven/Assets/StarterAssets/ThirdPersonController/Scripts/ThirdPersonController.cs:99]: little trick for future proofing: make sure you gate all your changes to Starter assets with something like // MTT CHANGE START and // MTT CHANGE STOP something like that. This way when merging new StarterAsset content later, you'll know which parts you did change yourself and which are just merge changes.
Hopefully once we'll update Starter assets with NGO we'll have less of those issues.

File: Basic/ClientDriven/Assets/UI Toolkit/2DSpaceShooterUIStyleSheet.uss
Comment (SamuelBellomo) [inline Basic/ClientDriven/Assets/UI Toolkit/2DSpaceShooterUIStyleSheet.uss:1]: I see 2D space shooter in the title, I'm assuming this was copied from the other bitesize?
I'm wondering if couldn't just create ourselves a common package between bitesize projects for things like these. Were there many changes to that sheet vs the space shooter?
might be something to sync with Jil

File: Basic/ClientDriven/Packages/manifest.json
Comment (SamuelBellomo) [inline Basic/ClientDriven/Packages/manifest.json:9]: adding a comment here to block this until it's merged boss room side.

File: Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs
Comment (fernando-cortez) [reply Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:56]: Right. If it makes sense to you, I've created a task to update ClientDriven doc & create a PR with the changes here: [MTT-5106](https://jira.unity3d.com/browse/MTT-5106)

File: Basic/ClientDriven/Assets/Scripts/ServerPlayerMove.cs
Comment (fernando-cortez) [reply Basic/ClientDriven/Assets/Scripts/ServerPlayerMove.cs:97]: When we were trying things out, the changes we made were client side on sending RPCs to the server. This modification on the server side of code resolved the issue. 
Though what's curious to me is moving this code to a coroutine and waiting for the next FixedUpdate did not produce the same results. I want to follow up with SDK on this offline, but this will absolutely go into the updated doc, as a new section.

File: Basic/ClientDriven/Packages/manifest.json
Comment (fernando-cortez) [reply Basic/ClientDriven/Packages/manifest.json:9]: I'll leave your comment open until the Boss Room update has been merged.

File: Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs
Comment (SamuelBellomo) [reply Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:56]: When is that fix gonna land? Cause as soon as we release this PR, docs will break.

File: Basic/ClientDriven/Assets/Scripts/ServerPlayerMove.cs
Comment (SamuelBellomo) [reply Basic/ClientDriven/Assets/Scripts/ServerPlayerMove.cs:97]: before telling our users they should do this, it'd be great if we understood exactly why (and make sure this isn't an issue with the SDK that we're just building a workaround for --> which would need to be reverted a month or two later...)

File: CHANGELOG.md
Comment (SamuelBellomo) [inline CHANGELOG.md:9]: It'd be great if we could discuss some of the bug fixes and improvements to Starter Assets (or link to the doc that does) that users would need to do on their side to make this work.

File: Basic/ClientDriven/Assets/Scripts/ServerPlayerMove.cs
Comment (fernando-cortez) [reply Basic/ClientDriven/Assets/Scripts/ServerPlayerMove.cs]: This has been since removed. The previous functionality of dropping an ingredient overhead was restored. This issue will be addressed (if need be) in the follow-up PR that integrates the pickup animations.

File: Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs
Comment (fernando-cortez) [reply Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:56]: Created this [docs PR](https://jira.unity3d.com/browse/MTT-5013) with scripting changes!

File: CHANGELOG.md
Comment (fernando-cortez) [reply CHANGELOG.md:9]: I've created a Jira task [here](https://jira.unity3d.com/browse/MTT-5179) to track this doc.

File: Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs
Comment (SamuelBellomo) [reply Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:56]: you mean this? https://github.com/Unity-Technologies/com.unity.multiplayer.docs/pull/865/files

File: Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs
Comment (SamuelBellomo) [inline Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:52]: mind adding a quick explanation on the why we did this? Those lines are part of the "magic" that makes Character Controller work, this should be highlighted.

File: Basic/ClientDriven/Assets/Prefabs/Ingredient.prefab
Comment (SamuelBellomo) [inline Basic/ClientDriven/Assets/Prefabs/Ingredient.prefab:655]: I can't remember, was looking up and down replicated as well? Or just character position and animations?

File: Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs
Comment (SamuelBellomo) [inline Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:40]:
```suggestion
        // ThirdPersonController & CharacterController are enabled only on owning clients. Ghost player objects have these two 
```

File: Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs
Comment (SamuelBellomo) [inline Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:42]: oh nice! good to know
@RikuTheFuffs @chrispope hopefully this will address most (?) issues raised by users on character controller usage? I'm not aware of other issues, if there's any other known issues, please let us know

File: Basic/ClientDriven/Assets/StarterAssets/ThirdPersonController/Scripts/ThirdPersonController.cs
Comment (SamuelBellomo) [inline Basic/ClientDriven/Assets/StarterAssets/ThirdPersonController/Scripts/ThirdPersonController.cs:103]: mmm so wait, doesn't that reduce the amount of functionality from the original Starter Assets? If users want to change the character's speed, with this change here the animations wouldn't react to it? Why aren't we trying to sync that magnitude info? Or deduce it from the character's current speed?

File: Basic/ClientDriven/Assets/Prefabs/Ingredient.prefab
Comment (fernando-cortez) [reply Basic/ClientDriven/Assets/Prefabs/Ingredient.prefab:655]: Synced offline. Will look into adding looking rotation to a follow-up update.

File: Basic/ClientDriven/Assets/StarterAssets/ThirdPersonController/Scripts/ThirdPersonController.cs
Comment (fernando-cortez) [reply Basic/ClientDriven/Assets/StarterAssets/ThirdPersonController/Scripts/ThirdPersonController.cs]: Followed up offline and this has since been re-enabled!

File: CHANGELOG.md
Comment (fernando-cortez) [reply CHANGELOG.md:9]: Closed the task, per offline discussion.

File: Basic/ClientDriven/Assets/Scripts/ServerIngredient.cs
Comment (RikuTheFuffs) [reply Basic/ClientDriven/Assets/Scripts/ServerIngredient.cs]: I remember somebody from the engine team pointing that out long time ago to me in a PR, but tbh I don't remember for which Unity version that observation was valid (and if it was by design or a bug of a specific major version)

Comment (RikuTheFuffs) [reply Basic/ClientDriven/Assets/Scripts/ServerIngredient.cs]: just re-asked [here](https://unity.slack.com/archives/C9SQHJGN6/p1669907293851319) to be sure

File: Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs
Comment (RikuTheFuffs) [reply Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:42]: Other issues I read about were related to the `PlayerInput` component being enabled by default, which caused its `Actions` field to be initialized with some weird new instance of `Input Action Asset`.
An easy solution to that is to disable the component by default, and to enable it only on the localPlayer

image

File: Basic/ClientDriven/Assets/Scripts/ServerIngredient.cs
Comment (RikuTheFuffs) [reply Basic/ClientDriven/Assets/Scripts/ServerIngredient.cs]: got an answer: "there's no impact in using both [SerializeField] + public". Sorry for the false alarm!

File: Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs
Comment (fernando-cortez) [reply Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:42]: That's a comment on proper usage, @RikuTheFuffs? Just checking since that component is disabled on the PlayerArmature_Networked player prefab, and only enabled on spawn for the owning player.

Comment (SamuelBellomo) [reply Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:42]: Might worth to still note this in our guide, to make sure inputs are handled only by owner clients and not other clients

**Review by SamuelBellomo — APPROVED** [review]

File: Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs
Comment (fernando-cortez) [reply Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:42]: Added [MTT-5190](https://jira.unity3d.com/browse/MTT-5190) as a follow up to the ClientDriven doc.

**Pull request merged**
